src/toil/fileStores/cachingFileStore.py

```diff
@@ -1589,16 +1588,83 @@ def _readGlobalFileWithCache(self, fileStoreID, localFilePath, symlink, readerID
                     # Wait for other people's downloads to progress.
                     time.sleep(self.contentionBackoff)
 
+    @contextmanager
+    def _with_copying_reference_to_upload(self, file_store_id: FileID, reader_id: str, local_file_path: Optional[str] = None) -> Generator:
+        """
+        Get a context manager that gives you either the local file path for a
+        copyuing reference to the given file, or None if that file is not in an
+        'uploadable' or 'uploading' state.
+
+        It is the caller's responsibility to actually do the copy, if they
+        intend to. The local file is not created.
+
+        If the reference remains a copying reference, it is removed at the end
+        of the context manager.
+
+        :param file_store_id: job store id for the file
+        :param str reader_id: Job ID of the job reading the file
+        :param str local_file_path: absolute destination path, if a particular one is desired
+        """
+
+        if not local_file_path:
+            # Generate a file path for the reference if one is not provided.
+            local_file_path = self.getLocalTempFileName()
+
+        # Try and make a 'copying' reference to such a file
+        self._write([('INSERT INTO refs SELECT ?, id, ?, ? FROM files WHERE id = ? AND (state = ? OR state = ?)',
```

Comment:
I'm not entirely familiar with the syntax here, so please ignore if this doesn't make sense, but could:

```FROM files WHERE id = ? AND (state = ? OR state = ?)```

be just:

```FROM files WHERE id = ? AND state = ?```

?

Reply:
No, I need to match either `uploadable` or `uploading`, which are the two distinct `state` values for which there's data that isn't fully written back yet, so I need an `OR` in the expression. Maybe if the state was really a set of flags we could make the condition shorter, but right now it isn't.